pkg/egresspolicy/egresspolicy.go

```diff
@@ -0,0 +1,520 @@
+		}).Error("failed to get valid pod metadata")
+		return
+	}
+
+	if policies, ok := manager.policyPods[id]; ok {
+		for _, policy := range policies {
+			config := manager.policyConfigs[policy]
+			manager.deleteEgressMap(config, podData)
+		}
+		delete(manager.policyPods, id)
+	}
+	delete(manager.podDataStore, id)
+}
+
+func (manager *Manager) getPodMetadata(pod *slimcorev1.Pod, podIPs []string) (*podMetadata, error) {
+	podId := k8s.ServiceID{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	namespace, err := manager.getNamespaceForPod(pod)
+	if namespace == nil || err != nil {
+		log.Warn("Failed in getting namespace")
+		data := &podMetadata{
+			ips:    podIPs,
+			labels: pod.GetLabels(),
+			id:     podId,
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (m *Manager) getNamespaceForPod(pod *slimcorev1.Pod) (*slimcorev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := m.storeGetter.GetStore("namespace")
+	nsItem, exists, err := nsStore.GetByKey(pod.GetNamespace())
+	if err != nil {
+		log.WithError(err).Warn("Failed to get namespace")
+		return nil, err
+	}
+	if !exists {
+		log.WithFields(logrus.Fields{"nsName": nsName}).Warn("Namespace does not exist")
+		return nil, nil
+	}
+	return nsItem.(*slimcorev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	return nil
+}
+
+// upsertPolicyPod updates or insert to pod policy mapping for given policy config and pods,
+// it also upserts the pod data catch and egress map to keep in sync
+func (manager *Manager) upsertPolicyPod(config *EgressPolicyConfig, pods ...*podMetadata) {
+	for _, pod := range pods {
+		podIp := net.ParseIP(pod.ips[0])
+		if podPolicies, ok := manager.policyPods[pod.id]; ok {
+			newPodPolicy := true
+			for _, poID := range podPolicies {
+				// Existing pod policy update, noop for this policy
+				if poID == config.id {
+					newPodPolicy = false
+					break
+				}
+			}
+			if newPodPolicy {
+				// Pod selected by a new policy
+				manager.policyPods[pod.id] = append(manager.policyPods[pod.id], config.id)
+				manager.updateEgressMap(podIp, config)
+				manager.podDataStore[pod.id] = pod
+			}
+		} else {
+			// Pod selected by a policy for the first time
+			pp := []policyID{config.id}
+			manager.policyPods[pod.id] = pp
+			manager.updateEgressMap(podIp, config)
+			manager.podDataStore[pod.id] = pod
+		}
+	}
+}
+
+func (manager *Manager) updateEgressMap(sip net.IP, config *EgressPolicyConfig) error {
+	var err error
+	for _, dstCIDR := range config.dstCIDRs {
+		//TODO validate fields before updating the map
```

Comment:
Ditto.

Reply:
This is resolved since we already validates the fields beforehand. So here we expect the config to be valid.
comment is now removed.